base_exception/models/base_exception.py

```diff
@@ -0,0 +1,254 @@
+        'exception.rule',
+        string='Exceptions')
+    ignore_exception = fields.Boolean('Ignore Exceptions', copy=False)
+
+    @api.depends('exception_ids', 'ignore_exception')
+    def _compute_main_error(self):
+        for obj in self:
+            if not obj.ignore_exception and obj.exception_ids:
+                obj.main_exception_id = obj.exception_ids[0]
+            else:
+                obj.main_exception_id = False
+
+    @api.multi
+    def _popup_exceptions(self):
+        action = self._get_popup_action()
+        action = action.read()[0]
+        action.update({
+            'context': {
+                'active_id': self.ids[0],
+                'active_ids': self.ids
+            }
+        })
+        return action
+
+    @api.model
+    def _get_popup_action(self):
+        action = self.env.ref('base_exception.action_exception_rule_confirm')
+        return action
+
+    @api.multi
+    def _check_exception(self):
+        """
+        This method must be used in a constraint that must be created in the
+        object that inherits for base.exception.
+        for sale :
+        @api.constrains('ignore_exception',)
+        def sale_check_exception(self):
+            ...
+            ...
+            self._check_exception
+        """
+        exception_ids = self.detect_exceptions()
+        if exception_ids:
+            exceptions = self.env['exception.rule'].browse(exception_ids)
+            raise ValidationError('\n'.join(exceptions.mapped('name')))
+
+    @api.multi
+    def test_exceptions(self):
+        """
+        Condition method for the workflow from draft to confirm
+        """
+        if self.detect_exceptions():
+            return False
+        return True
+
+    @api.multi
+    def detect_exceptions(self):
+        """returns the list of exception_ids for all the considered base.exceptions
+        """
+        if not self:
+            return []
+        exception_obj = self.env['exception.rule']
+        all_exceptions = exception_obj.sudo().search(
+            [('rule_group', '=', self[0].rule_group)])
+        model_exceptions = all_exceptions.filtered(
+            lambda ex: ex.model == self._name)
+        sub_exceptions = all_exceptions.filtered(
+            lambda ex: ex.model != self._name)
+
+        all_exception_ids = []
+        for obj in self:
+            if obj.ignore_exception:
+                continue
+            exception_ids = obj._detect_exceptions(
+                model_exceptions, sub_exceptions)
+            obj.exception_ids = [(6, 0, exception_ids)]
+            all_exception_ids += exception_ids
+        return all_exception_ids
+
+    @api.model
+    def _exception_rule_eval_context(self, obj_name, rec):
+        user = self.env['res.users'].browse(self._uid)
+        return {obj_name: rec,
+                'self': self.pool.get(rec._name),
+                'object': rec,
+                'obj': rec,
+                'pool': self.pool,
```

Comment:
old api ?

Reply:
No, this is not old API but variables which can use in Python code with safe_eval

Reply:
@mourad-ehm self.pool is certainly deprecated. You cannot use that.

I see it was migrated in v10 like that  but it's incorrect. Use self.env[rec._name] and 'env': self.env instead of pool.

But I think changing function to @api.multi and returning {'self': self} should be sufficent. So, you can change the function signature removing obj_name and rec parameters!